cypress/plugins/nodeManager/plugin.ts

```diff
@@ -0,0 +1,345 @@
+import * as path from "path";
+import * as childProcess from "child_process";
+import fetch from "node-fetch";
+import waitOn from "wait-on";
+import type { NodeSession } from "./shared";
+import { Commands } from "./shared";
+
+type NodeId = number;
+type PeerAddress = string;
+type AuthToken = string;
+
+const ROOT_PATH = path.join(__dirname, "../../../");
+
+// IP to which all started processes will bind to.
+const HOST = "127.0.0.1";
+
+const sleep = async (ms: number) => {
+  await new Promise(resolve => setTimeout(resolve, ms));
+};
+
+class Logger {
+  prefix: string;
+  indentationLevel: number;
+
+  constructor({ prefix = "", indentationLevel = 1 }) {
+    this.prefix = prefix;
+    this.indentationLevel = indentationLevel;
+  }
+
+  log(message: string) {
+    const indentation = " ".repeat(this.indentationLevel * 2);
+    console.log(indentation + this.prefix + message);
+  }
+}
+
+// Because it's not possible to mix tagged union types and extending
+// interfaces, we have to repeat the "inherited" attributes in each node state.
+interface ConfiguredNode {
+  kind: "configured";
```

Comment:
Not a blocker, but isn't common to create an enum for these `kind`s?

Reply:
Not in Typescript, if I understand the [docs](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#discriminating-unions) correctly:

> A common technique for working with unions is to have **a single field which uses literal types** which you can use to let TypeScript narrow down the possible current type.

Maybe @geigerzaehler knows more?

Reply:
Here's an example: https://github.com/radicle-dev/radicle-upstream/blob/master/ui/src/project.ts#L22-L36

Reply:
Is there any added benefit in wrapping the sate in an enum? Typescript already guards against using the wrong string.

![Screenshot 2021-01-28 at 14 23 13](https://user-images.githubusercontent.com/158411/106144378-67c50d80-6174-11eb-849a-4ae8a82571df.png)

![Screenshot 2021-01-28 at 14 21 44](https://user-images.githubusercontent.com/158411/106144271-4532f480-6174-11eb-898d-b404f2083d35.png)

It seems the enum version is slightly more verbose, and if there is no added benefit, I'd rather go with where I have to type less.

Reply:
I guess using enums is a convention in this codebase, so here it is: https://github.com/radicle-dev/radicle-upstream/pull/1492/commits/f9abb6d96640f4c930325d129e14ff730c331b82.

Reply:
Ya, the linter saves us, but having an `enum` reduces the cognitive load -- you can jump to the enum and see all the cases laid out for you :) Thanks for improving it :100: